### ISSUE TITLE
[Windows] Birdtray fails to start Thunderbird when Thunderbird updates

### ISSUE THREAD

**Issue by Abestanis** [gyunaev/birdtray#67]
On Windows, when BirdTray starts up and attempts to start up Thunderbird, it displays an error when Thunderbird is applying an upgrade.
BirdTray displays an error dialog stating:
```
Cannot start Thunderbird
Error starting Thunderbird as "...":
Process failed to start: The process cannot access the file because it is being used by another process.
```
Thunderbird will eventually start after applying the upgrade. BirdTray will not recognize that, until the error dialog is closed.

**Comment by gyunaev** [comment]
Probably should retry start in this case after a timeout?

**Comment by Abestanis** [comment]
I don't think simply retrying the start is going to work, because the same logic would also continuously try to start Thunderbird when the user configured an incorrect path to the Thunderbird executable.
Also, the updater automatically starts Thunderbird after the update. If we start an additional Thunderbird process, it will exit immediately as well, because Thunderbird prevents multiple instances from running simultaneously.
I also don't like the idea, that we could potentially start Thunderbird while the update is still in process, which could lead to errors during the installation and potentially broken installations.

The cleanest thing to to would be to supress the error if we detect the updater process, wait for the updater process to exit and then attach to the Thunderbird process that the updater spawned.

Unfortunately, there is no unified way to wait for a process that we didn't start directly in Qt, so this would require platform dependent code similar to the windowtools. This would also mean that we can no longer just check the existence of `mThunderbirdProcess`, since there is no way to generate a QProcess object for a process that we didn't create ourselves. We can still do that in the common case though, but we'd need to handle the update situation as a special case.

Waiting for the process is easy on Windows, but could be problematic on Linux, since I don't know if the updater process and the new Thunderbird process still count as child processes of the birdtray process.

The detection of the updater process can be done via the name, [they are probably final and won't change](https://dxr.mozilla.org/thunderbird/source/toolkit/xre/nsUpdateDriver.cpp#106).

## To summarize:
* On error, detect if the updater process is running. If not, show the error dialog, else
* wait for the updater process to exit.
* Then check for the Thunderbird process, if not found, try to start it again.
* If found, attach to the process so we get notified when it quits.

What are your thoughts on this?

**Comment by gyunaev** [comment]
This will only work on Windows. On Linux the updates are applied through the package manager, and this logic isn't needed. No idea about Mac.

**Comment by Abestanis** [comment]
I believe this will also work for Linux. While it is true that Thunderbird will usually be installed through the package manager on Linux, it is possible to [install it without a package manager](https://support.mozilla.org/en-US/kb/installing-thunderbird-linux#w_installing-outside-of-a-package-manager). I don't know how common that is, but it will also have this issue.

**Comment by gyunaev** [comment]
But there is no "file is being used by another process" kind of error on Linux; its not possible to open the file in SHARE_MODE_NONE here.

**Comment by Abestanis** [comment]
Hmm, you might be correct. I still want to test it though.

Did you get a chance to look at the pull request? I would like to get your feedback on it.

**Comment by gyunaev** [comment]
Yeah, but I'm not sure we need to bring in all that complexity to handle a rather simple isolated case. Wouldn't it be better to check the startup error code, and in case if it is "already used", suppress the error and simply attempt to restart again?

**Comment by Abestanis** [comment]
I admit the code is not complex than before, but for the usual case (no update) the code path stays the same (uses the QProcess).

As I said, I don't want us to start Thunderbird while the updater is still running. I don't want to be responsible for broken Thunderbird installations.

**Comment by Abestanis** [comment]
Also, I don't see a way with QProcess to differentiate between a `Not found` and `Already used` error besides the error message, but that is locale dependent.

**Comment by gyunaev** [comment]
Would switch then to use CreateProcess on Windows instead of QProcess.

**Comment by Abestanis** [comment]
That would be an option, but then it is also platform dependent and the code gets complicated. I don't see why that would be better then the code I propose in the pull request, which has more functionality.

**Comment by Abestanis** [comment]
> But there is no "file is being used by another process" kind of error on Linux; its not possible to open the file in SHARE_MODE_NONE here.

Ok, I managed to test it a few times on a Linux Vm and didn't notice any problems, so you are correct, Linux does not need any special handling. I also tested it a few times with the changes from the pull request to make sure it would not introduce any problems.

I want to test this on a Mac as well, but before I do that, I need you to tell me if you can live with the added complexity for the better user experience or if you won't accept the pull request and want me to find a simpler solution.

**Comment by gyunaev** [comment]
I'm really not sure about this one, as it brings one lots of code and introduces further complexity which need to be QAd (I'm especially concerned on Linux platform which doesn't need any of this).

And this fixes a really minor issue, which would happen three or so times a year. Honestly if you didn't work on this, I'd just close it as "wontfix". It doesn't really sound a big issue to close an error dialog three times a year. For those who do see this as a problem, maybe we should just disable "restart Thunderbird automatically" option on Windows?

**Comment by gyunaev** [comment]
Also the issue raised here seem to be the error dialog. So there are other possible solutions:

- Replace the error dialog with systray notification;
- Auto-close the error dialog in 5/10 seconds if the user didn't press Ok;

We can also attempt to open the thunderbird executable file for writing (if it doesn't open, the process is either running or being overwritten - again this only works on Windows).

**Comment by Abestanis** [comment]
Let me start by saying that I do understand your point. You are trying to keep the code complexity to a minimum to keep maintaining this project as easy as possible and that is a totally valid goal.

But please don't close this as wontfix when there is a solution to avoid the problem. Sure, this might just look like a small annoyance to you (although personally I hate it every time it happens), but you need to look at this from a perspective of an inexperienced user who is randomly confronted with an error dialog multiple times per year. This is not only a bad user experience, but you can be sure, that as Birdtray gets more popular for Windows users, there will be multiple bug reports about this per year.

So let's try to find a compromise. We should under NO circumstances open the Thunderbird executable for writing, to avoid messing with the installation process. I'm not even sure if we can get the permissions to open the exe for writing.

But I can revert the handling of the Thunderbird process back to how it was before and only use the Process handle to check and wait for the updater process on Windows. This would basically remove any changes for Linux/Mac and make the remaining changes Windows-only. Currently we don't really need to detect if the Thunderbird process finishes, because we get that information from the closing window event. The only reason I did it that way was to work towards a non-polling version of Birdtray, which would only wait on callbacks instead of checking for the Thunderbird window every 500 ms.

I will push a commit which implements the minimal changes to the pull request, so you can decide if I reduced the complexity enough. I think it will be a fair compromise between keeping the complexity low but also fixing this.

**Comment by Abestanis** [comment]
Thank you
